doc/designs/quic-design/quic-record-layer.md

```diff
@@ -0,0 +1,353 @@
+QUIC Record Layer API
+=====================
+
+The QUIC record layer (QRL) performs the following functions:
+
+- The handling of received datagrams and the initial processing of the
+  packets within them, including packet header processing and packet
+  decryption.
+
+- The serialization, encryption and transmission of sent packets.
+
+The QRL relies on the DEMUX, which is responsible for actual reads of datagrams
+from a datagram network BIO. The DEMUX performs a preliminary examination of
+received packet headers to determine a destination connection ID and passes
+received datagrams to a specific instance of a QUIC record layer based on the
+destination connection ID.
+
+The QRL does not process or examine decrypted packet payloads Processing of
+frames is performed by other components, which read them from the QRL.
+
+The QRL does not use the same method structures or polymorphic interface as the
+TLS record layer, as there is no benefit to doing so and the structure of the
+records returned differs substantially.
+
+Broadly, the interface to the QRL looks like this:
+
+Instantiation
+-------------
+
+```c
+#define OSSL_QUIC_ENC_LEVEL_INITIAL     0
+#define OSSL_QUIC_ENC_LEVEL_HANDSHAKE   1
+#define OSSL_QUIC_ENC_LEVEL_0RTT        2
+#define OSSL_QUIC_ENC_LEVEL_1RTT        3
+
+typedef struct ossl_qrl_st OSSL_QRL;
+
+typedef struct ossl_qrl_args_st {
+    OSSL_LIB_CTX *libctx;
+    const char   *propq;
+
+    /*
+     * This list is not exhaustive and will expand as the QRL
+     * is developed and its requirements become more clear.
+     */
+
+    /* Datagram network BIO to use. */
+    BIO    *rx_bio, *tx_bio;
+    /* Length of connection IDs in short packets in bytes. */
+    size_t  short_conn_id_len;
+    /* key ... */
+    /* IV ... */
+    /* cipher information ... */
+
+} OSSL_QRL_ARGS;
+
+/*
+ * Instantiates a new QRL. A pointer to the QRL is written
+ * to *qrl. Returns 1 on success or 0 on failure.
+ */
+int ossl_qrl_new(const OSSL_QRL_ARGS *args, OSSL_QRL **qrl);
```

Comment:
This deviates quite a lot from how we usually write constructors...  usually, we have them return the constructed object, or NULL on failure.